File: SignalServiceKit/src/Tests/TestKeychainStorage.m
//
//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
//

#import "TestKeychainStorage.h"
#import "NSData+OWS.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@implementation TestKeychainStorage

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return self;
    }

    self.dataMap = [NSMutableDictionary new];

    return self;
}

- (NSString *_Nullable)stringForKey:(NSString *)key
                            service:(NSString *)service
                              error:(NSError *_Nullable *_Nullable)error
{
    OWSAssert(error);
    OWSAssert(key.length > 0);
    OWSAssert(service.length > 0);

    *error = nil;

    NSString *mapKey = [NSString stringWithFormat:@"%@-%@", service, key];
    NSData *_Nullable data = self.dataMap[mapKey];
    if (!data) {
        NSLog(@"stringForKey:%@ service:%@ -> nil", key, service);
        return nil;
    }
    NSString *string = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    NSLog(@"stringForKey:%@ service:%@ -> %@", key, service, string);
    return string;
}

- (BOOL)setWithString:(NSString *)string
               forKey:(NSString *)key
              service:(NSString *)service
                error:(NSError *_Nullable *_Nullable)error
{
    OWSAssert(error);
    OWSAssert(key.length > 0);
    OWSAssert(service.length > 0);

    *error = nil;

    NSLog(@"setWithString:%@ service:%@ -> %@", key, service, string);

    NSString *mapKey = [NSString stringWithFormat:@"%@-%@", service, key];
    self.dataMap[mapKey] = [string dataUsingEncoding:NSUTF8StringEncoding];
    return YES;
}

- (NSData *_Nullable)dataForKey:(NSString *)key service:(NSString *)service error:(NSError *_Nullable *_Nullable)error
{
    OWSAssert(error);
    OWSAssert(key.length > 0);
    OWSAssert(service.length > 0);

    *error = nil;

    NSString *mapKey = [NSString stringWithFormat:@"%@-%@", service, key];
    NSData *_Nullable data = self.dataMap[mapKey];
    NSLog(@"dataForKey:%@ service:%@ -> %@", key, service, data.hexadecimalString);
    return data;
}

- (BOOL)setWithData:(NSData *)data
             forKey:(NSString *)key
            service:(NSString *)service
              error:(NSError *_Nullable *_Nullable)error
{
    OWSAssert(error);
    OWSAssert(key.length > 0);
    OWSAssert(service.length > 0);

    *error = nil;

    NSLog(@"setWithData:%@ service:%@ -> %@", key, service, data.hexadecimalString);

    NSString *mapKey = [NSString stringWithFormat:@"%@-%@", service, key];
    self.dataMap[mapKey] = data;
    return YES;
}

- (BOOL)removeWithKey:(NSString *)key service:(NSString *)service error:(NSError *_Nullable *_Nullable)error
{
    OWSAssert(error);
    OWSAssert(key.length > 0);
    OWSAssert(service.length > 0);

    *error = nil;

    NSLog(@"removeWithKey:%@ service:%@", key, service);

    NSString *mapKey = [NSString stringWithFormat:@"%@-%@", service, key];
    [self.dataMap removeObjectForKey:mapKey];
    return YES;
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Tests/TestKeychainStorage.swift
//
//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
//

import Foundation

#if DEBUG

@objc
public class SSKTestKeychainStorage: NSObject, SSKKeychainStorage {

    private var dataMap = [String: Data]()

    override init() {
        super.init()
    }

    @objc public func string(forService service: String, key: String) throws -> String {
        let data = try self.data(forService: service, key: key)
        guard let string = String(bytes: data, encoding: String.Encoding.utf8) else {
            throw KeychainStorageError.failure(description: "\(logTag) could not retrieve string")
        }
        return string
    }

    @objc public func set(string: String, service: String, key: String) throws {
        guard let data = string.data(using: String.Encoding.utf8) else {
            throw KeychainStorageError.failure(description: "\(logTag) could not store data")
        }
        try set(data: data, service: service, key: key)
    }

    private func key(forService service: String, key: String) -> String {
        return "\(service) \(key)"
    }

    @objc public func data(forService service: String, key: String) throws -> Data {
        let key = self.key(forService: service, key: key)
        guard let data = dataMap[key] else {
            throw KeychainStorageError.failure(description: "\(logTag) could not retrieve data")
        }
        return data
    }

    @objc public func set(data: Data, service: String, key: String) throws {
        let key = self.key(forService: service, key: key)
        dataMap[key] = data
    }

    @objc public func remove(service: String, key: String) throws {
        let key = self.key(forService: service, key: key)
        dataMap.removeValue(forKey: key)
    }
}

#endif
